kafka-streams-framework/src/main/java/org/hypertrace/core/kafkastreams/framework/punctuators/AbstractThrottledPunctuator.java

```diff
@@ -0,0 +1,97 @@
+package org.hypertrace.core.kafkastreams.framework.punctuators;
+
+import java.time.Clock;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Optional;
+import lombok.extern.slf4j.Slf4j;
+import org.apache.kafka.streams.KeyValue;
+import org.apache.kafka.streams.processor.Punctuator;
+import org.apache.kafka.streams.state.KeyValueIterator;
+import org.apache.kafka.streams.state.KeyValueStore;
+import org.hypertrace.core.kafkastreams.framework.punctuators.action.TaskResult;
+
+@Slf4j
+public abstract class AbstractThrottledPunctuator<T> implements Punctuator {
+  private final Clock clock;
+  private final KeyValueStore<Long, ArrayList<T>> objectStore;
+  private final ThrottledPunctuatorConfig config;
+
+  public AbstractThrottledPunctuator(
+      Clock clock,
+      ThrottledPunctuatorConfig config,
+      KeyValueStore<Long, ArrayList<T>> objectStore) {
+    this.clock = clock;
+    this.config = config;
+    this.objectStore = objectStore;
+  }
+
+  public void scheduleTask(long atTimestampInMs, T object) {
+    long windowAlignedTimestamp = getWindowAlignedTimestamp(atTimestampInMs);
+    ArrayList<T> objectsAtWindow =
+        Optional.ofNullable(objectStore.get(windowAlignedTimestamp)).orElse(new ArrayList<>());
+    objectsAtWindow.add(object);
+    objectStore.put(windowAlignedTimestamp, objectsAtWindow);
+  }
+
+  public boolean rescheduleTask(long atTimestampInMs, long toTimestampInMs, T object) {
+    if (cancelTask(atTimestampInMs, object)) {
+      scheduleTask(toTimestampInMs, object);
+      return true;
+    } else {
+      return false;
+    }
+  }
+
+  public boolean cancelTask(long atTimestampInMs, T object) {
+    long windowAlignedTimestamp = getWindowAlignedTimestamp(atTimestampInMs);
+    ArrayList<T> objectsAtWindow =
+        Optional.ofNullable(objectStore.get(windowAlignedTimestamp)).orElse(new ArrayList<>());
+    boolean removed = objectsAtWindow.remove(object);
```

Comment:
Agreed, but even say if we don't remove and track another variable say completed state for each object, that too isn't any better because we still have to write it back. Open to suggestions regarding making this better

Reply:
> cancellation is auto-handled when that key(windowAlignedTimestamp) is removed

But when is that condition triggered? Currently the way it is implemented key is removed only when the corresponding value list becomes empty

Reply:
Updated impl to be more optimal after discussion with Laxman, please take a look

Reply:
Lgtm